// bot/index.js
var request = require('request');

var data = {};

function Bot(key, mode, url) {
  this.key = key || 'default';
  this.mode = mode || 'training';
  this.url = url || 'http://s.smtcs.rocks:9989';
  this.data = {};
  console.log(this);
}

Bot.prototype.set = function(key, value) {
  Bot.prototype[name] = value;
};

Bot.prototype.unset = function(key) {
  delete Bot.prototype[key];
};

Bot.prototype.get = function(key) {
  return Bot.prototype[key];
};

Bot.prototype.start = function() {
  request.post({
    url: this.url + '/api' + this.mode,
    form: {
      key: this.key
    }
  }, function(err, res, body) {
    data = body;
    console.log(res.statusCode);
  });
  return this;
};

Bot.prototype.move = function(dir) {
  process.nextTick(function() {
    request.post({url: this.url + '/api/' + this.mode, form: {key: this.key, dir: dir}}, function(err, res, body) {
      // this.data = body;
      console.log(body);
      console.log('Moved', dir);
      // console.log(res.statusCode);
    });
  });
};

module.exports = Bot;
