// bot/index.js
var request = require('request');
var PF = require('pathfinding');
var Promise = require('bluebird');
var bot;

/*@Desc: Bot constructor.  Creates the bot object and sets starting information for the server. 
 *
 */
function Bot(key, mode, url) {
  this.key = key;
  this.mode = mode || 'training';
  this.serverURL = url === undefined ? 'http://vindinium.org' : url;
  this.URL = this.serverURL + '/api/' + this.mode;
  this.counter;
  this.gameStart = true;
  this.freeMines = [];
  this.bot1mines = [];
  this.bot2mines = [];
  this.bot3mines = [];
  this.bot4mines = [];
  this.bots = [];
  this.taverns = [];

  bot = this;
}

/*@Desc: Promise structure that runs each round of the game and then calls itself to continue the loop of playing.
 *
 */
Bot.prototype.runGame = function(){
 // console.log("promise sendREquest");
  bot.sendRequest().then(function(){
 //   console.log("promise parse");
    bot.parseTheData()}).then(function(){
 //     console.log("promise botBrain");
      bot.botBrain()}).then(function(){
 //       console.log("startingOver!");
        bot.runGame()}); 
};

/*@Desc: Sends post requests to the server.  Initiates the game, or if the game has already started, sends the player's move and gets
 *the next set of information to be parsed.
 *
 */
Bot.prototype.sendRequest = function (){
  var _this = bot;
  _this.mines = [];
  _this.bots = [];
  _this.taverns = [];
  return new Promise(function(resolve, reject){
    var _this = bot;
    console.log("play URL :" +  _this.URL +  " : " + _this.playURL);
    console.log("going to post first startgame:" + _this.URL + ", " + _this.key);    
    if(_this.gameStart === true){   
      _this.gameStart = false;
      request.post({url: _this.URL, form: {key: _this.key, map: "m1"}}, function(err, res, body) {
        _this.data = JSON.parse(body);
        console.log("This is view URL from first start: " + _this.data.viewUrl);
        _this.playURL = _this.data.playUrl;
        console.log("this is playURL and playUrl" + _this.data.playUrl + ":" + _this.playURL);
        resolve();
      })
    }
    else{
      if(_this.data.game.finished !== true){
        console.log("requestion! " + _this.goDir);
        request.post({url: _this.data.playUrl, form: {key: _this.key, dir: _this.goDir}}, function(err, res, body){
   //       console.log("doh" + _this.data.viewUrl);
          _this.data = JSON.parse(body);
          resolve(); 
        })
      }
    }
  }) 
};

/*@Desc: Each of the characters organized into their own objects with position, health, mines owned
 *health = current health
 */
Bot.prototype.parseTheData = function (){
 return new Promise(function(resolve, reject){ 
  var _this = bot;
  
  _this.bot1 = {
    id: _this.data.game.heroes[0].id,
    pos:{
      x: _this.data.game.heroes[0].pos.y,
      y: _this.data.game.heroes[0].pos.x,
    },
    life: _this.data.game.heroes[0].life,
    mineCount: _this.data.game.heroes[0].mineCount
  };
  _this.bot2 = {
    id: _this.data.game.heroes[1].id,
    pos:{
      x: _this.data.game.heroes[1].pos.y,
      y: _this.data.game.heroes[1].pos.x,
    },
    life: _this.data.game.heroes[1].life,
    mineCount: _this.data.game.heroes[1].mineCount
  };
  _this.bot3 = {
    id: _this.data.game.heroes[2].id,
    pos:{
      x: _this.data.game.heroes[2].pos.y,
      y: _this.data.game.heroes[2].pos.x,
    },
    life: _this.data.game.heroes[2].life,
    mineCount: _this.data.game.heroes[2].mineCount
  };
  _this.bot4 = {
    id: _this.data.game.heroes[3],
    pos:{
      x: _this.data.game.heroes[3].pos.y,
      y: _this.data.game.heroes[3].pos.x,
    },
    life: _this.data.game.heroes[3].life,
    mineCount: _this.data.game.heroes[3].mineCount
  };
  _this.yourBot = {
  };
  if(_this.data.hero.id === 1){_this.yourBot = _this.bot1}
  if(_this.data.hero.id === 2){_this.yourBot = _this.bot2}
  if(_this.data.hero.id === 3){_this.yourBot = _this.bot3}
  if(_this.data.hero.id === 4){_this.yourBot = _this.bot4}

  /* @DescRepackaging the bot.data so that it is more useful for the students.
   *
   * Bot.data currently holds the information from the JSON file in an object.
   *
   * botN will be an object with the botN's position and health
   *
   * map will be an array with the data for the map.
   *  " " = Walkable
   *  "#" = wall
   *  1 = hero 1
   *  2 = hero 2
   *  3 = hero 3
   *  4 = hero 4
   *  "t" = Tavern
   *  "m" = unoned Mine
   *  "a" "b" "c" "d" = Player Owned Mine by player 1, 2, 3, 4 respectively
   */


  var tempMap = _this.data.game.board.tiles;
  var tempMapSize = _this.data.game.board.size;
  _this.map = [];
  for(var i = 0; i < tempMapSize; i++){
    _this.map.push([]);
  }
  for(var i = 0; i < tempMapSize; i++){
    for(var j = 0; j < tempMapSize; j++){
      _this.map[i].push([]);
    }
  }

  var counter = 0;
  var xPos = 0;
  var yPos = 0;
  for(var i = 0; i < (tempMapSize * tempMapSize * 2); i+=2){
    yPos = counter;
    if(tempMap[i] === "#") {_this.map[xPos][yPos] = "#";}
    if(tempMap[i] === " ") {_this.map[xPos][yPos] = " ";}
    if(tempMap[i] === "[") {_this.map[xPos][yPos] = "t"; _this.taverns.push([yPos, xPos]);}
    if(tempMap[i] === "@") {
      _this.map[xPos][yPos] = tempMap[i+1]; _this.bots.push([yPos, xPos]);
    }
    if(tempMap[i] === "$") {
      if(tempMap[i+1] === "1"){
        _this.map[xPos][yPos] = "a"; _this.bot1mines.push([yPos, xPos]);
      }
      else if(tempMap[i+1] === "2"){
        _this.map[xPos][yPos] = "b"; _this.bot2mines.push([yPos, xPos]);
      }
      else if(tempMap[i+1] === "3"){
        _this.map[xPos][yPos] = "c"; _this.bot3mines.push([yPos, xPos]);
      }
      else if(tempMap[i+1] === "4"){
        _this.map[xPos][yPos] = "d"; _this.bot4mines.push([yPos, xPos]);
      }
      else{
        _this.map[xPos][yPos] = "m"; _this.freeMines.push([yPos, xPos]);
      }
    }
    counter++;
    if(counter === tempMapSize){
      counter = 0;
      xPos++;
    }
  }
  console.log(bot.map);
  resolve();  
})};


/*@Desc: Creates a bot.pathfinder method for users that will take the location of two items on the map and return the direction
 * that the chracter should go in order to reach the second object based on A* pathfinding
 *@Params - two arrays - each array must have two ints, equal to the X, Y coordinates on the map. Should be  equal to the [x][y] in the map object.
 */
Bot.prototype.findPath = function(pos1, pos2){
    var _this = bot;
    //console.log("pathfinder bot :" + bot);
    var pathFinderMap = _this.map;
    var gridLength = _this.data.game.board.size;
    
    // Changes all walkables into 0s and non walkables into 1s
    for(var i = 0; i < gridLength; i++){
      for(var j = 0; j < gridLength; j++){
        if(_this.map[i][j] === ' '){pathFinderMap[i][j] = 0}
        else{pathFinderMap[i][j] = 1};
      }
    }
    
    console.log(pathFinderMap);

    var grid = new PF.Grid(gridLength, gridLength, pathFinderMap);
    grid.setWalkableAt(pos1[0], pos1[1], true);
    grid.setWalkableAt(pos2[0], pos2[1], true);

    //console.log(pos1[0], pos1[1], pos2[0], pos2[1]);
    var finder = new PF.AStarFinder();
    
    var path = finder.findPath(pos1[0], pos1[1], pos2[0], pos2[1], grid);
    
      console.log("length of path: " + path.length);
      //console.log("next spot!" + path[1][0] + ", "+ path[1][1]);
      if (path === null) {
        console.log("none found +(");
        _this.pathDir = "none";  
      }
      else {
        if(path[1][0] === pos1[0]){
          if(path[1][1] < pos1[1]){
            console.log("set to north");
            _this.pathDir = "north";
            return "north";
          }
          else{
            console.log("set to south");
            _this.pathDir = "south";
            return "south";
          }
        }
        else{
          if(path[1][0] > pos1[0]){
            console.log("set to east");
            _this.pathDir = "east";
            return "east";
          }
          else{
            console.log("set to west");
            _this.pathDir = "west";
            return "west";
          }
         }
        }
  
   console.log(path);
    console.log("woop");
    return _this.pathDir;
    };
  

/*@Desc: Bot.findDistance method that finds the number of steps that would be required to go from one spot to another
 *@Param: two arrays - each array has the x,y spot for one position
 *@Return: one int - the int is equal to the number of steps to get from one spot to another
 */
Bot.prototype.findDistance = function(pos1, pos2){
  var _this = bot;
  var pathFinderMap = _this.map;
  var gridLength = _this.data.game.board.size;
  
  // Changes all walkables into 0s and non walkables into 1s
  for(var i = 0; i < gridLength; i++){
    for(var j = 0; j < gridLength; j++){
      if(_this.map[i][j] === ' '){pathFinderMap[i][j] = 0}
      else{pathFinderMap[i][j] = 1};
    }
  }
  //sets the grid and makes the start and end positions walkable
  var grid = new PF.Grid(gridLength, gridLength, pathFinderMap);
  grid.setWalkableAt(pos1[0], pos1[1], true);
  grid.setWalkableAt(pos2[0], pos2[1], true);

    //console.log(pos1[0], pos1[1], pos2[0], pos2[1]);
  var finder = new PF.AStarFinder();
  var path = finder.findPath(pos1[0], pos1[1], pos2[0], pos2[1], grid);
  if(path === []){return "no path";}
  else {return path.length - 2;}
}

module.exports = Bot;
